src/ctia/domain/id.clj

```diff
@@ -0,0 +1,109 @@
+(ns ctia.domain.id
+  (:require [ctia.lib.url :as url])
+  (:import java.lang.String))
+
+(def short-id-re
+  (re-pattern (str "(" url/url-chars-re ")")))
+
+(def url-re
+  #"(https?):\/\/([-\da-zA-Z][-\da-zA-Z.]*)(:(\d+))?((\/[-\w.]+)*)\/ctia\/([a-z]+)\/")
+
+(def long-id-re
+  (re-pattern
+   (str url-re
+        short-id-re)))
+
+(defn make-long-id-str
+  [{:keys [hostname short-id path-prefix port protocol type]
+    :as parts}]
+  (when parts
+    (str protocol
+         "://"
+         hostname
+         (if port (str ":" port))
+         path-prefix
+         "/ctia/"
+         type
+         "/"
+         (url/encode short-id))))
+
+(defn long-id?
+  "Make an educated guess if this is a long-form ID"
+  [^String id]
+  ;; TODO - More specific checking?
+  (boolean (.startsWith id "http")))
+
+(def short-id?
+  "Make an educated guess if this is a short-form ID"
+  (complement long-id?))
+
+(defprotocol ID
+  (localize [this url-params]
+    "Localize this ID")
+  (short-id [this]
+    "The short part of the ID")
+  (long-id [this] [this url-params]
+    "Convert this record to a long ID string"))
+
+(defrecord CtiaId [hostname short-id path-prefix port protocol type]
+  ID
+  (localize [this url-show-params]
+    (merge this
+           (-> url-show-params
+               (update :path-prefix not-empty))))
```

Comment:
Is this planning to have multiple update/assoc/dissoc in future? If not, then the threading macro isn't needed